### PR TITLE
test: add linter for manager api

### PR DESCRIPTION
**Pull request by nic-chen**
add linter for manager api 

and fix lint error

close: #649

### CODE REVIEW

**Comment by nic-chen** [comment]
@spacewander  please take a look. thanks.

File: .github/workflows/golang-lint.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: golangci-lint
```
Comment (juzhiyuan) [inline .github/workflows/golang-lint.yml:1]: How about `golang-linter`?

File: .github/workflows/golang-lint.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
+jobs:
+  golangci:
+    name: lint
```
Comment (juzhiyuan) [inline .github/workflows/golang-lint.yml:11]: Run lint

File: .github/workflows/golang-lint.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: lint
+        run: |
+          make golangci-lint
```
Comment (juzhiyuan) [inline .github/workflows/golang-lint.yml:18]: There is no need to use `|` for 1 command.

File: api/internal/core/store/store_test.go [hunk 4]
```diff
@@ -580,7 +580,9 @@ func TestGenericStore_Create(t *testing.T) {
 			createCalled = true
 			assert.Equal(t, tc.wantKey, args[1], tc.caseDesc)
 			input := TestStruct{}
-			_ = json.Unmarshal([]byte(args[2].(string)), &input)
+			if err := json.Unmarshal([]byte(args[2].(string)), &input); err != nil {
+				panic(err)
```
Comment (juzhiyuan) [inline api/internal/core/store/store_test.go:584]: image

Not sure if this format is ok 🤔

Comment (juzhiyuan) [reply api/internal/core/store/store_test.go:584]: Kindly resolve this conversation if this format if ok.

File: api/internal/core/store/validate.go [hunk 5]
```diff
@@ -88,8 +88,9 @@ func NewAPISIXJsonSchemaValidator(jsonPath string) (Validator, error) {
 }
 
 func getPlugins(reqBody interface{}) map[string]interface{} {
-	switch reqBody.(type) {
+	switch bodyType := reqBody.(type) {
 	case *entity.Route:
+		fmt.Printf("type of reqBody: %#v", bodyType)
```
Comment (spacewander) [inline api/internal/core/store/validate.go:93]: Is the `fmt.Printf` for debug? Would be better if we can use log instead?

File: api/internal/handler/route/route.go [hunk 6]
```diff
@@ -320,7 +320,12 @@ func (h *Handler) BatchDelete(c droplet.Context) (interface{}, error) {
 	}
 
 	//delete stored script
-	h.scriptStore.BatchDelete(c.Context(), strings.Split(input.IDs, ","))
+	if err := h.scriptStore.BatchDelete(c.Context(), strings.Split(input.IDs, ",")); err != nil {
+		//try again
```
Comment (spacewander) [inline api/internal/handler/route/route.go:324]: Better to add a warning log when we have to try again.

File: Makefile [hunk 7]
```diff
@@ -22,4 +22,13 @@ ifeq ("$(wildcard .actions/openwhisk-utilities/scancode/scanCode.py)", "")
 	git clone https://github.com/apache/openwhisk-utilities.git .actions/openwhisk-utilities
 	cp .actions/ASF* .actions/openwhisk-utilities/scancode/
 endif
-	.actions/openwhisk-utilities/scancode/scanCode.py --config .actions/ASF-Release.cfg ./
\ No newline at end of file
+	.actions/openwhisk-utilities/scancode/scanCode.py --config .actions/ASF-Release.cfg ./
+
```
Comment (spacewander) [inline Makefile:26]: Need to add `SHELL := bash` as `command` is bash's builtin command.

**Review by spacewander — CHANGES_REQUESTED** [review]

File: .github/workflows/golang-lint.yml (hunk 1 above)
Comment (membphis) [reply .github/workflows/golang-lint.yml:1]: agree +1

simpler and clearer

File: api/conf/conf.go [hunk 8]
```diff
@@ -36,8 +36,8 @@ const (
 	EnvDEV   = "dev"
 	EnvLOCAL = "local"
 
-	confPath   = "/go/manager-api/conf.json"
-	schemaPath = "/go/manager-api/schema.json"
+	confJsonPath = "/go/manager-api/conf.json"
```
Comment (membphis) [inline api/conf/conf.go:39]: One PR for one thing, is it related to the PR?

File: api/internal/core/store/store_test.go (hunk 4 above)
Comment (gxthrj) [reply api/internal/core/store/store_test.go:584]: use `assert` instead of `panic`?

Comment (gxthrj) [inline api/internal/core/store/store_test.go:584]: use `assert`  instead of `panic`?

File: api/conf/conf.go (hunk 8 above)
Comment (nic-chen) [reply api/conf/conf.go:39]: fix lint err.

**Review by spacewander — APPROVED** [review]

File: .github/workflows/golang-lint.yml [hunk 9]
```diff
@@ -0,0 +1,16 @@
+name: golang-lint
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
+jobs:
```
Comment (juzhiyuan) [inline .github/workflows/golang-lint.yml:9]: extra line is needed

File: .github/workflows/golang-lint.yml [hunk 10]
```diff
@@ -0,0 +1,16 @@
+name: golang-lint
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
```
Comment (juzhiyuan) [inline .github/workflows/golang-lint.yml:8]: Have run on 2.0 branch?

File: api/internal/handler/upstream/upstream_test.go [hunk 11]
```diff
@@ -46,52 +46,53 @@ func TestUpstream(t *testing.T) {
...
+		              "http_statuses": [200, 201],
+		              "successes": 3
+		          },
+		          "unhealthy": {
+		              "http_statuses": [500],
+		              "http_failures": 3,
+		              "tcp_failures": 3
+		          }
+		      }
+		  }
```
Comment (juzhiyuan) [inline api/internal/handler/upstream/upstream_test.go:92]: image

What happened here?

**Review by juzhiyuan — APPROVED** [review]
Just a better-change comment

**Review by juzhiyuan — APPROVED** [review]

File: api/conf/conf.go (hunk 8 above)
Comment (moonming) [inline api/conf/conf.go:39]: Is this a hard-coded configuration file?

**Review by moonming — CHANGES_REQUESTED** [review]

File: api/conf/conf.go (hunk 8 above)
Comment (juzhiyuan) [reply api/conf/conf.go:39]: No, could be overridden by conf file.

Comment (juzhiyuan) [reply api/conf/conf.go:39]: Sorry for my typo, this path would be overridden by ENV.

https://github.com/nic-chen/incubator-apisix-dashboard/blob/go-lint/docs/deploy.md#note

image

Comment (juzhiyuan) [reply api/conf/conf.go:39]: And @nic-chen  will resolve this in the future for easy to use

Comment (moonming) [reply api/conf/conf.go:39]: hemm, anyway, we can not hard-coded in code base.

Comment (nic-chen) [reply api/conf/conf.go:39]: @moonming 
it's a history issue, we have a plan to refactor the config part in `2.1`

File: api/internal/handler/upstream/upstream_test.go [hunk 12]
```diff
@@ -51,45 +51,45 @@ func TestUpstream(t *testing.T) {
 		"description": "upstream upstream",
 		"type": "roundrobin",
 		"nodes": [{
-		          "host": "a.a.com",
-		          "port": 80,
-		          "weight": 1
-		      }],
+			"host": "a.a.com",
```
Comment (membphis) [inline api/internal/handler/upstream/upstream_test.go:54]: do we found those problems with the `linter` tools?

Comment (membphis) [reply api/internal/handler/upstream/upstream_test.go:54]: @nic-chen any news?

File: api/filter/logging.go [hunk 13]
```diff
@@ -42,16 +43,17 @@ func RequestLogHandler() gin.HandlerFunc {
 
 		param, _ := c.Get("requestBody")
 
-		switch param.(type) {
+		switch paramType := param.(type) {
 		case []byte:
 			param = string(param.([]byte))
+			fmt.Printf("type of param: %#v", paramType)
```
Comment (membphis) [inline api/filter/logging.go:49]: change to `log.info`

File: api/internal/core/store/validate.go [hunk 14]
```diff
@@ -88,14 +89,17 @@ func NewAPISIXJsonSchemaValidator(jsonPath string) (Validator, error) {
 }
 
 func getPlugins(reqBody interface{}) map[string]interface{} {
-	switch reqBody.(type) {
+	switch bodyType := reqBody.(type) {
 	case *entity.Route:
+		log.Printf("type of reqBody: %#v", bodyType)
```
Comment (membphis) [inline api/internal/core/store/validate.go:94]: use `log.infof` is better

File: api/internal/handler/upstream/upstream_test.go (hunk 12 above)
Comment (membphis) [reply api/internal/handler/upstream/upstream_test.go:54]: I got it, we need to process this by handle

**Review by membphis — APPROVED** [review]
LGTM, except minor points. After fixing them, we can merge this PR.

File: api/filter/logging.go (hunk 13 above)
Comment (nic-chen) [reply api/filter/logging.go]: Sorry, the original log library does not support info, warn, etc.. levels, need to submit another PR to do this.

Comment (nic-chen) [reply api/filter/logging.go]: I change to log first.

File: api/internal/core/store/validate.go (hunk 14 above)
Comment (nic-chen) [reply api/internal/core/store/validate.go:94]: Sorry, the original log library does not support info, warn, etc.. levels, need to submit another PR to do this.

File: api/filter/logging.go [hunk 15]
```diff
@@ -46,7 +46,7 @@ func RequestLogHandler() gin.HandlerFunc {
 		switch paramType := param.(type) {
 		case []byte:
 			param = string(param.([]byte))
-			fmt.Printf("type of param: %#v", paramType)
+			log.Printf("type of param: %#v", paramType)
```
Comment (membphis) [inline api/filter/logging.go:49]: should we use `log.info` here?

File: api/internal/core/store/validate.go [hunk 16]
```diff
@@ -194,7 +194,7 @@ func checkConf(reqBody interface{}) error {
 	switch bodyType := reqBody.(type) {
 	case *entity.Route:
 		route := reqBody.(*entity.Route)
-		fmt.Printf("type of reqBody: %#v", bodyType)
+		log.Printf("type of reqBody: %#v", bodyType)
```
Comment (membphis) [inline api/internal/core/store/validate.go:197]: ditto

**Review by membphis — CHANGES_REQUESTED** [review]

File: api/filter/logging.go (hunk 15 above)
Comment (nic-chen) [reply api/filter/logging.go]: OK.

File: .github/workflows/golang-lint.yml [hunk 17]
```diff
@@ -0,0 +1,19 @@
+name: golang-lint
+on:
+  push:
+    branches:
+      - master
```
Comment (juzhiyuan) [inline .github/workflows/golang-lint.yml:5]: trigger 2.0

**Comment by juzhiyuan** [comment]
Ready to merge?

**Comment by juzhiyuan** [comment]
> LGTM, except minor points. After fixing them, we can merge this PR.

Those minor points get updated, once all Actions pass, I will merge this.

**Pull request merged**